## Algorithm の選択

In [1]:
# 余分な Warning を非表示にする
import warnings

warnings.filterwarnings('ignore')

# Library の Import
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# matplotlib 日本語化対応
import japanize_matplotlib

# 表示 Option の調整
np.set_printoptions(suppress=True, precision=4)  # NumPy の浮動小数点の表示精度
pd.options.display.float_format = '{:.4f}'.format  # pandas での浮動小数点の表示精度
pd.set_option('display.max_columns', None)  # DataFrame ですべての項目を表示
plt.rcParams['font.size'] = 14  # Graph の Default font 指定
random_seed = 123  # 乱数の種

In [2]:
# Sample data の読み込み
# （乳がん疾患 Data）

# Data の load
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

# 入力 Data: x(30次元)
# 正解 Data: y
x = cancer.data
y = cancer.target

# Sample data の分割

# Data 分割の Parameters
test_size = 0.1

# Data 分割
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=test_size, random_state=random_seed, stratify=y)

# 分割 Size 確認
print(x.shape)
print(x_train.shape)
print(x_test.shape)

(569, 30)
(512, 30)
(57, 30)


## 複数 Algorithm の List を作成

In [3]:
# 複数 Algorithm で精度を比較
# 結果が同じになるよう random_state は同一にする

# 線形回帰
from sklearn.linear_model import LogisticRegression

algorithm = LogisticRegression(random_state=random_seed)

# Support vector machine(Kernel)
from sklearn.svm import SVC

algorithm2 = SVC(kernel='rbf', random_state=random_seed)

# 決定木
from sklearn.tree import DecisionTreeClassifier

algorithm3 = DecisionTreeClassifier(random_state=random_seed)

# Random forest
from sklearn.ensemble import RandomForestClassifier

algorithm4 = RandomForestClassifier(random_state=random_seed)

# XGBoost
from xgboost import XGBClassifier

algorithm5 = XGBClassifier(random_state=random_seed)

# Algorithm の List 作成
algorithms = [algorithm, algorithm2, algorithm3, algorithm4, algorithm5]

In [4]:
# 複数 Algorithm で精度比較
for algorithm in algorithms:
    # 訓練 Data で学習
    algorithm.fit(x_train, y_train)
    # 検証 Data で精度測定
    score = algorithm.score(x_test, y_test)
    # Algorithm 名取得
    name = algorithm.__class__.__name__
    # 精度と Algorithm 名表示
    print(f'score: {score:.4f} {name}')

score: 0.9474 LogisticRegression
score: 0.8947 SVC
score: 0.9474 DecisionTreeClassifier
score: 0.9298 RandomForestClassifier
[05:56:13] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
score: 0.9825 XGBClassifier


- Support Vector Machine は、精度が悪い
- XGBoost は精度が良く検証 Data に対して98.25% となっている
- 他の３つの候補は似たり寄ったりだが、Logistic 回帰と決定木の精度が Random Forest よりは良い。

この見た目の結論は、本当に正しいのか？